### ISSUE TITLE
ReflectionAllocator is broken on JDK 16

### ISSUE THREAD

**Issue by Ali-RS** [jMonkeyEngine/jmonkeyengine#1674]
`SEVERE: Buffer cannot be destroyed: java.nio.DirectFloatBufferU`

Even by adding this JVM arg (which used to solve that on jdk 11+)

"--add-opens=java.base/jdk.internal.ref=ALL-UNNAMED"

you can test it with

https://github.com/jMonkeyEngine/jmonkeyengine/blob/d76d4dc8c67bc99c5fd197d41a102639194a4e3b/jme3-examples/src/main/java/jme3test/app/TestReleaseDirectMemory.java#L43

Forum post:
https://hub.jmonkeyengine.org/t/jme-on-jdk-16/44411/6?u=ali_rs

**Comment by Ali-RS** [comment]
We might consider deprecating ReflectionAllocator for the 3.5 release.

**Comment by stephengold** [comment]
Deprecating `ReflectionAllocator` is another idea that should be discussed at the Forum/Hub ASAP.

**Comment by Ali-RS** [comment]
Sorry @stephengold, I thought I have deleted that comment. I guess it's to late to delete it now but I will modify it to sense a bit nice;)

You are right deperecation is another topic that should be discoussed on forum.

Regards

**Comment by stephengold** [comment]
Here's the diagnostic message I got when I ran `TestReleaseDirectMemory` on 64-bit Linux using OpenJDK 16.0.1: 
```text
Dec 08, 2021 12:55:58 PM com.jme3.system.JmeDesktopSystem initialize
INFO: Running on jMonkeyEngine 3.5.0-SNAPSHOT
 * Branch: master
 * Git Hash: 645b1bb
 * Build Date: 2021-12-08
Inconsistency detected by ld.so: dl-lookup.c: 111: check_match: Assertion `version->filename == NULL || ! _dl_name_match_p (version->filename, map)' failed!
```

**Comment by stephengold** [comment]
Using OpenJDK 14.0.2, I get additional diagnostic messages:
```text
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by com.jme3.util.ReflectionAllocator (file:/home/sgold/Git/jmonkeyengine/jme3-core/build/libs/jme3-core-3.5.0-SNAPSHOT.jar) to method sun.nio.ch.DirectBuffer.cleaner()
WARNING: Please consider reporting this to the maintainers of com.jme3.util.ReflectionAllocator
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
Dec 08, 2021 1:08:29 PM com.jme3.system.JmeDesktopSystem initialize
INFO: Running on jMonkeyEngine 3.5.0-SNAPSHOT
 * Branch: master
 * Git Hash: 645b1bb
 * Build Date: 2021-12-08
Inconsistency detected by ld.so: dl-lookup.c: 111: check_match: Assertion `version->filename == NULL || ! _dl_name_match_p (version->filename, map)' failed!
```

After adding the "--add-opens=java.base/jdk.internal.ref=ALL-UNNAMED" JVM argument, the test still crashes similarly with OpenJDK 14.0.2 and even with OpenJDK 11.0.11 . So I don't think this crash has much to do with JDK 16.

**Comment by stephengold** [comment]
@Ali-RS: I need some help reproducing the "Buffer cannot be destroyed" crash. So far, I haven't even managed to **create** a buffer with anything other than OpenJDK 8.

**Comment by Sailsman63** [comment]
>Inconsistency detected by ld.so: dl-lookup.c: 111: check_match: Assertion `version->filename == NULL || ! _dl_name_match_p (version->filename, map)' failed!

This looks very familiar. This error is typical of binary incompatibility between native libraries and newer builds of the JRE. Last time this came up, I think the only solution was to use lwjgl3, which was built against the newer runtimes.

There may be more details on the hub.

**Comment by Ali-RS** [comment]
@stephengold have you tried with AdoptOpenJDK?

**Comment by danielperano** [comment]
Chiming in on @stephengold's request.

I'm running on an Oracle build of JDK 16 with LWJGL 3 with no issues.

**Comment by Ali-RS** [comment]
@stephengold were you able to reproduce the "Buffer cannot be destroyed" crash?

**Comment by stephengold** [comment]
I never got that far.

**Comment by Ali-RS** [comment]
A note if you use LWJGL3:

LWJGL3 will **statically** register its own buffer allocator at LwjglContext:

https://github.com/jMonkeyEngine/jmonkeyengine/blob/ceb356462aeedf7769dced52e1d5572b01deca31/jme3-lwjgl3/src/main/java/com/jme3/system/lwjgl/LwjglContext.java#L93-L104

which will be **statically** initialized at this line into the **final** field `allocator`:


https://github.com/jMonkeyEngine/jmonkeyengine/blob/ceb356462aeedf7769dced52e1d5572b01deca31/jme3-core/src/main/java/com/jme3/util/BufferUtils.java#L65

So you must make sure to register `ReflectionAllocator` before the static block run, in my case I specify it as JVM argument in Gradle build:

'-Dcom.jme3.BufferAllocatorImplementation=com.jme3.util.ReflectionAllocator'

* **Off-topic:**
Sometimes the order of these static executions is different, e.g the one in the BufferUtils runs before the LwjglContext, which will cause `BufferAllocator allocator` to instantiate into `Reflectionallocator` and bypass the LWJGL3 allocator.  :open_mouth:

**Comment by riccardobl** [comment]
The ReflectionAllocator was an hack to begin with. Does anyone know if it is used outside of the lwjgl2 renderer? 
If not i assume it will be deprecated with lwjgl2 at some point.

> Sometimes the order of these static executions is different, e.g the one in the BufferUtils runs before the LwjglContext, which will cause BufferAllocator allocator to get initialized with Reflectionallocator and bypass the LWJGL3 allocator. open_mouth
So best to specify it via JVM args.

It seems we should find a different way to set the allocator, maybe the renderers should have the buffer allocator under the same class name and namespace so that when the modules are switched it gets replaced without setting it "manually".

**Comment by Ali-RS** [comment]
>Does anyone know if it is used outside of the lwjgl2 renderer?

Yes, `AndroidBufferAllocator` uses it. It is error-prone on android as well. See https://hub.jmonkeyengine.org/t/androidbufferallocator-and-nonsdkapiusedviolation/45124/2?u=ali_rs

**Comment by riccardobl** [comment]
We already have the jme3-android-native module, the JNI allocator can be added there. 
Using PrimitiveAllocator might be problematic especially with low memory devices.

**Comment by riccardobl** [comment]
> It seems we should find a different way to set the allocator, maybe the renderers should have the buffer allocator under the same class name and namespace so that when the modules are switched it gets replaced without setting it "manually".

To elaborate on this:

The first part of my proposal is to have a NativeAllocator that is implemented by every rendered module internally (eg. lwjgl3 will use its allocator, lwjgl2 will use ReflectionAllocator, android will use the jni allocator) with the same class on the same path: com.jme3.util.NativeAllocator . 
In this way when a different renderer module is used its native allocator is loaded.

The next step of the solution is to have BufferAllocatorFactory load the allocator that is  specified with the property PROPERTY_BUFFER_ALLOCATOR_IMPLEMENTATION like it does now, but the default will always be NativeAllocator, if the specified class doesn't exist it will fallback to PrimitiveAllocator.


That should solve or provide a path to solve every possible issue. What do you think?

**Comment by Ali-RS** [comment]
Is `NativeAllocator` an interface? or a class that implements the `BufferAllocator` interface?

**Comment by riccardobl** [comment]
A class that implements BufferAllocator

**Comment by Ali-RS** [comment]
Ok, that sounds good to me. However, should note that Java9 module system does not permit split packages. So if at somepoint we want to switch to moudle system it will not work.

**Comment by Ali-RS** [comment]
Thank you @Scrappers-glitch

**Comment by Scrappers-glitch** [comment]
> Thank you @Scrappers-glitch

Btw, android internals use the same maneuver under `android.hardware.HardwareBuffer`, so we are on the right path :-) : 

The native binding for allocating mem on cpu/gpu : 
https://developer.android.com/reference/android/hardware/HardwareBuffer

The native lib : 
https://developer.android.com/ndk/reference/group/a-hardware-buffer

Sources at Android cs : 
https://cs.android.com/android/platform/superproject/+/master:external/mesa3d/include/android_stub/android/hardware_buffer.h;l=46?q=hardware_buffer.h&ss=android

**Comment by Scrappers-glitch** [comment]
@Ali-RS This may be why you have experienced this bug **on JDK-16** and it's absent on other jdks, `ReflectionAllocator` makes a direct call to `sun.*` packages, "Internal method signature are subject to change across jdks" : 
https://www.oracle.com/java/technologies/faq-sun-packages.html

**Comment by Sailsman63** [comment]
And in particular, there has been a serious push to remove the `sun.*` packages altogether.

However, if that were the cause you should be getting `java.lang.NoClassDefFoundError`, unless it's being swallowed somewhere.

**Comment by Scrappers-glitch** [comment]
Or guarded against reflective calls...i don't know why generally, but on android they made java reflection and invocation api a black list.

**Comment by Sailsman63** [comment]
> i don't know why generally

Mostly security. Reflection allows access to APIs that may not be safe for general code to use. Even desktop Java is moving to block reflection to a lot of internal packages. But in that case, you should be getting `IllegalReflectiveAccess` errors. Again, unless they are being swallowed by code.

**Comment by Scrappers-glitch** [comment]
I am done with Android Native buffer allocator, but there is something which is inconsistent, on `com.jme3.utl` package in the module jme3-android, you will find similar classes on jme3-core but designed for android but prefixed by `Android` which is not the case on other packages like `com.jme3.input` and `com.jme3.system`, they have their classes inside `com.jme3,input.android` and `com.jme3.system.android`, so i know this is another issue but what should i do about the Android native buffer ? Should i just name it `AndroidNativeBuffer` or should create a package android under `com.jme3.util` ?

**Comment by Ali-RS** [comment]
I thibk based on riccardobl idea:

>   It seems we should find a different way to set the allocator, maybe the renderers should have the buffer allocator under the same class name and namespace so that when the modules are switched it gets replaced without setting it "manually".

>To elaborate on this:

>The first part of my proposal is to have a NativeAllocator that is implemented by every rendered module internally (eg. lwjgl3 will use its allocator, lwjgl2 will use ReflectionAllocator, android will use the jni allocator) with the same class on the same path: com.jme3.util.NativeAllocator .
In this way when a different renderer module is used its native allocator is loaded.

I suppose we should put it under the `com.jme3.util` package and the name it `NativeAllocator` or maybe better to name it `NativeBufferAllocator`.

**Comment by Scrappers-glitch** [comment]
Would it create a conflict when we build the `NativeBufferAllocator` on desktop because both are at the same package `com.jme3.util` ?

**Comment by Scrappers-glitch** [comment]
However, other android packages like `com.jme3.input` and system has a sub folder android and the classes inside aren't prefixed by `Android`.

**Comment by Ali-RS** [comment]
>Would it create a conflict when we build the NativeBufferAllocator on desktop because both are at the same package com.jme3.util ?

No, because at the runtime only one module (lwjgl or android) will be in classpath. On desktop only jme3-lwjgl module will be loaded and on android only jme3-android module will be loaded.

**Comment by Ali-RS** [comment]
By the way, I am not sure if riccardobl approach is the best way or we can have even better approaches. But this is another topic and we need to discuss this on the forum. 

I am ok with either of these namings:

`com.jme3.util.AndroidNativeBufferAllocator`

or

`com.jme3.util.NativeBufferAllocator`

**Comment by Scrappers-glitch** [comment]
> I am not sure if riccardobl approach is the best way or if we can use a better alternative. But this is another topic and we need to discuss this on the forum.

I guess its the only way, I read some links of lwjgl describing different approaches, by far handling direct buffers natively is the best way, and btw `PrimitiveAllocator` might be also problematic on desktop arm devices as Raspberry PIs (kinda the same memory problems as android), so I think its better also to implement the approach on desktop instead of ReflectionAllocator.

**Comment by Ali-RS** [comment]
Note, I am talking about this issue:

>Sometimes the order of these static executions is different, e.g the one in the BufferUtils runs before the LwjglContext, which will cause BufferAllocator allocator to get initialized with Reflectionallocator and bypass the LWJGL3 allocator. open_mouth
    So best to specify it via JVM args.

riccardobl sugested to use a different approach to solve the issue of staticly setting allocator in code :

>It seems we should find a different way to set the allocator, maybe the renderers should have the buffer allocator under the same class name and namespace so that when the modules are switched it gets replaced without setting it "manually".

maybe an enhancement to this would be to config the allocator name via properties file (for each backend) instead of using a fixed class name. 

So that BufferAllocatorFactory will read the property  PROPERTY_BUFFER_ALLOCATOR_IMPLEMENTATION from the properties file.

**Comment by Scrappers-glitch** [comment]
Oh, it's better to infer this conversation to the forum, so i haven't looked up on `LwjglAllocator` yet, but this solution can be an alternative for natives on desktop ?

**Comment by Ali-RS** [comment]
> but this solution can be an alternative for natives on desktop ?

Sorry not sure if I understand what you mean. `LwjglBufferAllocator` in jme3-lwjgl3 is already using native methods (via JNI through lwjgl3) and we are not going to replace it.

But if you are talking about lwjgl2, it does not use a native allocator, it uses RefectionAllocator, so yes we can write a native allocator for lwjgl2 to replace the RefectionAllocator, just like what you did for android.

**Comment by Scrappers-glitch** [comment]
Thanks for elaborating this, I will take a deep look into both `jme3-lwjgl` and `jme3-lwjgl3` modules, may be I can help in taking decisions, once I understand what's going on there, I will open a forum discussion yo help deciding what is best.

**Comment by Scrappers-glitch** [comment]
Further discussion on forums : 
https://hub.jmonkeyengine.org/t/directbufferallocator/45591

**Comment by Scrappers-glitch** [comment]
There is a new report today about this issue : 
https://hub.jmonkeyengine.org/t/illegal-reflective-access-warning/45603?u=pavl_g

**Comment by Ali-RS** [comment]
After diving more and adding some println to the suppressed exceptions I gathered some more info about error:

```
java.lang.reflect.InaccessibleObjectException: Unable to make public abstract jdk.internal.ref.Cleaner sun.nio.ch.DirectBuffer.cleaner() accessible: module java.base does not "exports sun.nio.ch" to unnamed module @4bca8b50
	at java.base/java.lang.reflect.AccessibleObject.checkCanSetAccessible(AccessibleObject.java:354)
	at java.base/java.lang.reflect.AccessibleObject.checkCanSetAccessible(AccessibleObject.java:297)
	at java.base/java.lang.reflect.Method.checkCanSetAccessible(Method.java:199)
	at java.base/java.lang.reflect.Method.setAccessible(Method.java:193)
	at com.jme3.util.ReflectionAllocator.loadMethod(ReflectionAllocator.java:84)
	at com.jme3.util.ReflectionAllocator.<clinit>(ReflectionAllocator.java:63)
	at java.base/java.lang.Class.forName0(Native Method)
	at java.base/java.lang.Class.forName(Class.java:375)
	at com.jme3.util.BufferAllocatorFactory.create(BufferAllocatorFactory.java:30)
	at com.jme3.util.BufferUtils.<clinit>(BufferUtils.java:65)
	at com.jme3.renderer.lwjgl.LwjglGL.<init>(LwjglGL.java:15)
	at com.jme3.system.lwjgl.LwjglContext.initContext(LwjglContext.java:283)
	at com.jme3.system.lwjgl.LwjglContext.initContextFirstTime(LwjglContext.java:265)
	at com.jme3.system.lwjgl.LwjglContext.internalCreate(LwjglContext.java:454)
	at com.jme3.system.lwjgl.LwjglAbstractDisplay.initInThread(LwjglAbstractDisplay.java:124)
	at com.jme3.system.lwjgl.LwjglAbstractDisplay.run(LwjglAbstractDisplay.java:221)
	at java.base/java.lang.Thread.run(Thread.java:833)
java.lang.ClassNotFoundException: sun.misc.Cleaner
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:641)
	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:188)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:520)
	at java.base/java.lang.Class.forName0(Native Method)
	at java.base/java.lang.Class.forName(Class.java:375)
	at com.jme3.util.ReflectionAllocator.loadMethod(ReflectionAllocator.java:83)
	at com.jme3.util.ReflectionAllocator.<clinit>(ReflectionAllocator.java:64)
	at java.base/java.lang.Class.forName0(Native Method)
	at java.base/java.lang.Class.forName(Class.java:375)
	at com.jme3.util.BufferAllocatorFactory.create(BufferAllocatorFactory.java:30)
	at com.jme3.util.BufferUtils.<clinit>(BufferUtils.java:65)
	at com.jme3.renderer.lwjgl.LwjglGL.<init>(LwjglGL.java:15)
	at com.jme3.system.lwjgl.LwjglContext.initContext(LwjglContext.java:283)
	at com.jme3.system.lwjgl.LwjglContext.initContextFirstTime(LwjglContext.java:265)
	at com.jme3.system.lwjgl.LwjglContext.internalCreate(LwjglContext.java:454)
	at com.jme3.system.lwjgl.LwjglAbstractDisplay.initInThread(LwjglAbstractDisplay.java:124)
	at com.jme3.system.lwjgl.LwjglAbstractDisplay.run(LwjglAbstractDisplay.java:221)
	at java.base/java.lang.Thread.run(Thread.java:833)
java.lang.NoSuchMethodException: sun.nio.ch.DirectBuffer.viewedBuffer()
	at java.base/java.lang.Class.getMethod(Class.java:2227)
	at com.jme3.util.ReflectionAllocator.loadMethod(ReflectionAllocator.java:83)
	at com.jme3.util.ReflectionAllocator.<clinit>(ReflectionAllocator.java:65)
	at java.base/java.lang.Class.forName0(Native Method)
	at java.base/java.lang.Class.forName(Class.java:375)
	at com.jme3.util.BufferAllocatorFactory.create(BufferAllocatorFactory.java:30)
	at com.jme3.util.BufferUtils.<clinit>(BufferUtils.java:65)
	at com.jme3.renderer.lwjgl.LwjglGL.<init>(LwjglGL.java:15)
	at com.jme3.system.lwjgl.LwjglContext.initContext(LwjglContext.java:283)
	at com.jme3.system.lwjgl.LwjglContext.initContextFirstTime(LwjglContext.java:265)
	at com.jme3.system.lwjgl.LwjglContext.internalCreate(LwjglContext.java:454)
	at com.jme3.system.lwjgl.LwjglAbstractDisplay.initInThread(LwjglAbstractDisplay.java:124)
	at com.jme3.system.lwjgl.LwjglAbstractDisplay.run(LwjglAbstractDisplay.java:221)
	at java.base/java.lang.Thread.run(Thread.java:833)
java.lang.reflect.InaccessibleObjectException: Unable to make public abstract java.lang.Object sun.nio.ch.DirectBuffer.attachment() accessible: module java.base does not "exports sun.nio.ch" to unnamed module @4bca8b50
	at java.base/java.lang.reflect.AccessibleObject.checkCanSetAccessible(AccessibleObject.java:354)
	at java.base/java.lang.reflect.AccessibleObject.checkCanSetAccessible(AccessibleObject.java:297)
	at java.base/java.lang.reflect.Method.checkCanSetAccessible(Method.java:199)
	at java.base/java.lang.reflect.Method.setAccessible(Method.java:193)
	at com.jme3.util.ReflectionAllocator.loadMethod(ReflectionAllocator.java:84)
	at com.jme3.util.ReflectionAllocator.<clinit>(ReflectionAllocator.java:68)
	at java.base/java.lang.Class.forName0(Native Method)
	at java.base/java.lang.Class.forName(Class.java:375)
	at com.jme3.util.BufferAllocatorFactory.create(BufferAllocatorFactory.java:30)
	at com.jme3.util.BufferUtils.<clinit>(BufferUtils.java:65)
	at com.jme3.renderer.lwjgl.LwjglGL.<init>(LwjglGL.java:15)
	at com.jme3.system.lwjgl.LwjglContext.initContext(LwjglContext.java:283)
	at com.jme3.system.lwjgl.LwjglContext.initContextFirstTime(LwjglContext.java:265)
	at com.jme3.system.lwjgl.LwjglContext.internalCreate(LwjglContext.java:454)
	at com.jme3.system.lwjgl.LwjglAbstractDisplay.initInThread(LwjglAbstractDisplay.java:124)
	at com.jme3.system.lwjgl.LwjglAbstractDisplay.run(LwjglAbstractDisplay.java:221)
```

I found out on java 16+ in addition to 
```
"--add-opens=java.base/jdk.internal.ref=ALL-UNNAMED" 
```
I must also add 
```
"--add-exports=java.base/sun.nio.ch=ALL-UNNAMED"
```
 then error `SEVERE: Buffer cannot be destroyed: java.nio.DirectFloatBufferU` will go away.

**Comment by Ali-RS** [comment]
>we should deprecate the ReflectionAllocator and replace it with a native one (like we did on android), if this is the case, let's continue our discussion on the forum before opening another issue.

I have already made a forum post regarding the deprecation of ReflectionAllocator:

https://hub.jmonkeyengine.org/t/deprecating-reflectionallocator/46350

>replace it with a native one (like we did on android)

That would be cool, are you interested in working on this?

**Comment by Scrappers-glitch** [comment]
Yeah, I will be happy to work on this, but I think if Lwjgl already (which I think it really does) provides a java implementation for a native allocator, we might use it.

**Comment by MeFisto94** [comment]
I agree with Pavl, that's just workarounds and ultimately this may stop working, the JRE already moved from a warning to an error. 
So we'd need to find different allocators. Without knowing what it's exact use is, there are probably a lot of different allocators to choose from, so we should do the research work here

**Comment by Ali-RS** [comment]
>but I think if Lwjgl already (which I think it really does) provides a java implementation for a native allocator, we might use it.

I had a quick glance at the lwjgl2 source code below, there are already a BufferUtils and MemoryUtils there but I could not find a destroy method,

https://github.com/jMonkeyEngine/lwjgl2

**Comment by Ali-RS** [comment]
>So we'd need to find different allocators. Without knowing what it's exact use is, there are probably a lot of different allocators to choose from, so we should do the research work here

I think for android, there was a discussion around that and calloc is suggested by core devs.

https://github.com/jMonkeyEngine/jmonkeyengine/pull/1821#discussion_r878656286

**Comment by Scrappers-glitch** [comment]
We should track lwjgl2 native header files to decide whether there is already a destroy criteria there (it must be there), however I cannot find any header file inside lwjgl2-jme ? I hope they are not inside the shared object files (which is a bad practice)....I will try to track them down...

**Comment by Scrappers-glitch** [comment]
> > So we'd need to find different allocators. Without knowing what it's exact use is, there are probably a lot of different allocators to choose from, so we should do the research work here
> 
> I think for android, there was a discussion around that, and "calloc" is suggested by the core team.
> 
> [#1821 (comment)](https://github.com/jMonkeyEngine/jmonkeyengine/pull/1821#discussion_r878656286)

Yes, because calloc = malloc + evaluating the buffer to NULL or zero.

**Comment by Ali-RS** [comment]
@Scrappers-glitch I think we better create it separate from the lwjgl2 project. I guess you might have a hard time tackling with building lwjgl2 natives for different OS.  Our lwjgl2 build uses precompiled natives.

**Comment by Scrappers-glitch** [comment]
> @Scrappers-glitch I think we better create it separate from the lwjgl2 project. (e.g. in a `jme3-calloc` project hosted on JME Github organization) You can copy whatever code you need from lwjgl2. I guess you might have a hard time tackling with lwjgl2 native builds for different OS. Our lwjgl2 build workflow uses precompiled natives.
> 
> I would like to know @stephengold opinion about this as well.

Looks clean, i don't mind to create a new API, however i am not so far good at building natives for cross-platform (i only do linux arches and android), so this might take a lot of time, i am currently learning cmake if it would bring something new to the engine, then let me know...

**Comment by Ali-RS** [comment]
I do have any experience building native code so I can not give any advise, but I guess our code team might suggest using Gradle and GitHub actions for this.  
@MeFisto94 and @stephengold might have a better opinion on the build toolset.

**Comment by Scrappers-glitch** [comment]
CMake is just a simple build script that you can use to generate native system build files and it's callable from gradle task with `Runtime.getRuntime().exec()` could be used to call the cmake script to build a shared object file before including it in the output `jme-alloc.jar`, I am not yet sure about how other architectures should be handled, you may give me some time to model this.

**Comment by Scrappers-glitch** [comment]
Once all architectures are built into `jme-alloc.jar`, other tasks like uploading to sonatype will be simple...

**Comment by Ali-RS** [comment]
No problem, take your time and ask on the forum if you need help.

**Comment by MeFisto94** [comment]
Before we build our own native dependency to only call `malloc` we should really just use one of the widely available allocators already.

One that is available to jwgl3 is jemalloc: https://javadoc.lwjgl.org/org/lwjgl/system/jemalloc/JEmalloc.html
Also considering that the only reason we have the ReflectionAllocator is that we don't want to wait for the direct buffer to be freed by the Garbage Collector (which is because an access to an already freed byte buffer may be catastrophic).
This is still a very active topic in the java world (https://stackoverflow.com/questions/3496508/deallocating-direct-buffer-native-memory-in-java-for-jogl)

Especially the hint to project panama is interesting, if you have a completely new API also supporting try-with-resources. 
Quoting https://openjdk.org/jeps/370:
> Moreover, working with direct buffers can be cumbersome since deallocation of the memory associated with direct buffers is left to the garbage collector; that is, only after a direct buffer is deemed unreachable by the garbage collector can the associated memory be released.
[...]
> While using JNI to access memory is also a possibility, the inherent costs associated with this solution make it seldom applicable in practice. The overall development pipeline is complex, since JNI requires the developer to write and maintain snippets of C code. JNI is also inherently slow, since a Java-to-native transition is required for each access.

So in general, I can't recommend anything right now, but if our buffers are mainly to communicate with openGL, memory mapping with a new API may be interesting. And a specialized allocator may make a difference, too. With malloc I don't know how thread safety looks like etc and you'd still need to get the native address of the byte buffer and try to wrap it etc. 

Furthermore, I wouldn't recommend cmake for something so simple, since you need visual studio and the cmake workload on windows, on mac os probably XCode (my mac can't even use any recent XCode anymore because it's "old") etc.
If it only targets CI, may even cross compile on linux (or have very simple manual gcc calls), _or_ try out the Gradle Support for C++ projects, as seems common on Android anyway.

**Comment by Scrappers-glitch** [comment]
@MeFisto94 Yes, you are right, we shouldn't create a new API for just malloc(), if using jemalloc will be enough then let me know.

About multi-threading and thread safety in jni: 
- You can attach the current native thread to the jvm.
- Or use the thread safe jvm pointer instead as the following: 
https://github.com/Software-Hardware-Codesign/Serial4j-v1/blob/4d52d4683d87b6e644dcc3b1c6535c18d1bb6928/src/main/native/include/JniUtils.h#L75-L96

I think i didn't respect this point while building android native buffer so far, so thank you for pointing this out, i totally forgot.

**Comment by Ali-RS** [comment]
>if using jemalloc will be enough then let me know.

I think it is ok if we can use lwjgl3 jemalloc lib.

**Comment by Scrappers-glitch** [comment]
> > if using jemalloc will be enough then let me know.
> 
> I am okay using lwjgl3 jemalloc lib. Can we use it without depending on lwjgl3-core?

I think it cannot, this is a script generated by [lwjgl.org/customize](https://www.lwjgl.org/customize).

```gradle
import org.gradle.internal.os.OperatingSystem

project.ext.lwjglVersion = "3.3.1"

switch (OperatingSystem.current()) {
	case OperatingSystem.LINUX:
		def osArch = System.getProperty("os.arch")
		project.ext.lwjglNatives = osArch.startsWith("arm") || osArch.startsWith("aarch64")
			? "natives-linux-${osArch.contains("64") || osArch.startsWith("armv8") ? "arm64" : "arm32"}"
			: "natives-linux"
		break
	case OperatingSystem.MAC_OS:
		project.ext.lwjglNatives = System.getProperty("os.arch").startsWith("aarch64") ? "natives-macos-arm64" : "natives-macos"
		break
	case OperatingSystem.WINDOWS:
		def osArch = System.getProperty("os.arch")
		project.ext.lwjglNatives = osArch.contains("64")
			? "natives-windows${osArch.startsWith("aarch64") ? "-arm64" : ""}"
			: "natives-windows-x86"
		break
}

repositories {
	mavenCentral()
}

dependencies {
	implementation platform("org.lwjgl:lwjgl-bom:$lwjglVersion")

	implementation "org.lwjgl:lwjgl"
	implementation "org.lwjgl:lwjgl-jemalloc"
	runtimeOnly "org.lwjgl:lwjgl::$lwjglNatives"
	runtimeOnly "org.lwjgl:lwjgl-jemalloc::$lwjglNatives"
}
```

**Comment by Scrappers-glitch** [comment]
Now, i am totally sure it cannot be used without lwjgl core, it utilizes the lwjgl core system package:
https://github.com/LWJGL/lwjgl3/blob/ee2dc3bbdc8d601428c1108cf8486e2aaf18fd07/modules/lwjgl/jemalloc/src/generated/java/org/lwjgl/system/jemalloc/JEmalloc.java#L12-L20

**Comment by Ali-RS** [comment]
If you think it is possible to rip only the required java code + natives from lwjgl3 then go that way but if you think creating our own thin API around malloc would be easier, then do it. You can use whatever build tool you are familiar with it does not matter.

What matters is that we want a `jme3-alloc.jar` with natives inside and a java class that implements `BufferAllocator`:

https://github.com/jMonkeyEngine/jmonkeyengine/blob/4db02fae9dd8ebb213b3a06316a53019ef9b620f/jme3-core/src/main/java/com/jme3/util/BufferAllocator.java#L6-L24

The jar should be hosted on maven central and then we will add a dependency to it in `jme3-lwjgl`.

**Comment by Scrappers-glitch** [comment]
> if you think it would be easy to use `GetDirectBufferAddress` from lwjgl2 code to get java ByteBufer memory address and destroy it with `stdlib` then use that.

I realized that. the code we implemented on android is really doing this indeed :-), but this option renders the same solution as our own API, we will have to recompile sources.

It's not possible to separate jemalloc from lwjgl3-core and lwjgl3-natives.

**Comment by MeFisto94** [comment]
In theory, an allocator is the backbone of an application, so don't over-rush this please.
As to thread safety, e.g, can you safely call free() on a memory block that has been malloc'ed() on another thread?
Can you free() a memory block that has _not_ been malloced (e.g. JVM's direct buffers could use the platform native API such as HeapAlloc on Windows).

Tuning the allocator is also yet another topic.
Also, if jemalloc doesn't work, there are hundreds of other allocators. I really like Netty's ByteBuf, e.g., but it's not API compatible with Java Standard ByteBuffers.

**Comment by Scrappers-glitch** [comment]
> As to thread safety, e.g, can you safely call free() on a memory block that has been malloc'ed() on another thread?
> Can you free() a memory block that has _not_ been malloced (e.g. JVM's direct buffers could use the platform native API such as HeapAlloc on Windows).

AFAIK, malloc() allocates new memory on the native heap which is visible to the jvm, use `JavaVM` instead of local `JNIEnv` to call the `GetAddress` method locally inside the native method.

**Comment by pspeed42** [comment]
Visible but not managed by.  I don't know if this is really a concern or not.

One of the big problems with Java6 was having to specify a huge direct heap because GC would free direct memory but did not consider it when deciding to run GC.  There is a concern that if we move to memory that Java does not manage that maybe we return to those awful days.  That would be unfortunate.

I admit I have not looked into what's going on here, though.  Not really something to take lightly in the alpha phase of a piece of software maybe.

**Comment by stephengold** [comment]
> can you safely call free() on a memory block that has been malloc'ed() on another thread?

Any modern C (or C++) compiler should provide a runtime library with thread-safe malloc()/free().

**Comment by Scrappers-glitch** [comment]
Okay, i think we are left with an additional option which is creating a new minimalistic allocation api (`BufferUtils` for example) on top of either [Libbulletjme](https://github.com/stephengold/Libbulletjme) hosted by @stephengold.

@stephengold Let me know your opinion about this option if this is feasible, and also if this will cause any conflict for people who still use other bullet physics libraries, if we are good with this option, assign me to this and i will work on a PR to the libbulletjme to add the BufferUtils and another PR to the engine to utilize a new `NativeDesktopAllocator`.

**Comment by stephengold** [comment]
+ Libbulletjme is a huge amount of code (253K lines of source, 370KB class JAR + a 3 KB native for each platform). I sincerely doubt JMonkeyEngine wants it as a dependency.
+ Minie and Libbulletjme define many of the same Java classes. Adding Libbulletjme as a dependency of JMonkeyEngine would break Minie in ways that would be difficult to fix.
+ While Libbulletjme implements automatic deallocation of (native) physics objects, it doesn't do so for direct buffers. For buffers, it uses a copy of JMonkeyEngine's `PrimitiveAllocator`---it never deallocates direct buffers! (I should fix that ASAP.)

I'm willing to create a new library for buffer allocation and deallocation, using a weak-reference tracking mechanism similar to what Libbulletjme uses for physics objects. However, I don't have a clear sense of how long it would take, and I'm unclear how much time remains before the JME 3.6 code freeze.

**Comment by Ali-RS** [comment]
>and I'm unclear how much time remains before the JME 3.6 code freeze.

It can be a separate project independent of JME releases.

**Comment by Scrappers-glitch** [comment]
Okay, i have a raw Algorithm library [arithmos](https://github.com/Arithmos-Algorithms/Arithmos), it's a standalone project with no dependencies, i will just refactor it to use any of [gradle-jni-templates](https://github.com/Software-Hardware-Codesign/gradle-jni-example) to get myself familiar with this until we reach a solution, i guess what will take most of the time (if not all) is building to windows and Mac systems, otherwise coding will not be a very hard task.

**Comment by Ali-RS** [comment]
Note that the `jme3-alloc` library should only contain the code required for buffer allocation and deallocation. It should not contains other stuff related to physics, image loading,...

**Comment by stephengold** [comment]
But first, I'm unclear why we even need `ReflectionAllocator`. What are the circumstances under which one cannot use `PrimitiveAllocator`?

**Comment by Scrappers-glitch** [comment]
May be using low-memory desktop devices like raspberries or Risc-V MCU ? It may give us some memory performance.

**Comment by stephengold** [comment]
Related discussion at the Discourse Forum/Hub:  https://hub.jmonkeyengine.org/t/jme-alloc-project/46356

**Comment by Scrappers-glitch** [comment]
> May be using low-memory desktop devices like raspberries or Risc-V MCU ? It may give us some memory performance.

I was wrong !

This project aims to provide a backward compatible direct memory allocation api for desktop on lwjgl-2.

Currently, I managed to build x86-64 binaries for Linux, Mac and windows.

I cannot build x86 binaries; because github-runners don't provide x86 systems, so there are 2 options, either to use another CI to build the other variants or create workarounds by downloading their toolchains in their respective systems (so for example downloading GCC for x86 Intel or looking around GCC options to provide a backward compatibility for the output binary).

**Comment by stephengold** [comment]
@riccardobl Why was this issue closed?

**Comment by riccardobl** [comment]
The ReflectionAllocator is not used by the engine anymore and [jme3-saferallocator](https://github.com/jMonkeyEngine/jmonkeyengine/tree/master/jme3-saferallocator) is basically the `jme3-alloc`  proposed here.
